### PR TITLE
Fix issue  149 - add some CI workflows

### PR DESCRIPTION
**Pull request by obnoxxx**
# Describe what this PR does #

This PR adds some github workflows to the CI as requested in #149 

## Is there anything that requires special attention ##

* Note that for mod check, this does not only add a workflow yaml file but also a `mod.check` make 
target to be easily invoked locally by developers
* The changes are split into multiple commits: one per item. 
 
Do you have any questions?

no questiions at this time.

Is the change backward compatible?

It should be backward compatible.

Provide any external context for the change, if any.

none.

## Related issues ##

Mention any github issues relevant to this PR. Adding below line
will help to auto close the issue once the PR is merged.

Fixes: #149 

## Future concerns ##

no future concerns currently. 

**Checklist:**

* [ ] **Commit Message Formatting**: Commit titles and messages follow
  guidelines in the [developer
  guide](https://github.com/ceph/ceph-csi-operator/blob/devel/docs/development-guide.md#commit-messages).
* [ ] Reviewed the developer guide on [Submitting a Pull
  Request](https://github.com/ceph/ceph-csi-operator/blob/devel/docs/development-guide.md#development-workflow)
* [ ] [Pending release
  notes](https://github.com/ceph/ceph-csi-operator/blob/devel/PendingReleaseNotes.md)
  updated with breaking and/or notable changes for the next major release.
* [ ] Documentation has been updated, if necessary.
* [ ] Unit tests have been added, if necessary.
* [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Comment by obnoxxx** [comment]
This is marked as draft since it does not cover all points of #149 yet.

**Comment by obnoxxx** [comment]
@Madhu-1  what remains is to add the stale workflow.

Any special requirements for this one or is it OK to use the rook workflow as a start?

https://github.com/rook/rook/blob/master/.github/workflows/stale.yaml

**Comment by parth-gr** [comment]
add some CI worjflows -> add some CI workflows

**Comment by obnoxxx** [comment]
@parth-gr wrote:
> add some CI worjflows -> add some CI workflows

good catch, thanks! fixed.

**Comment by obnoxxx** [comment]
replying to myself:
> @Madhu-1 what remains is to add the stale workflow.
> 
> Any special requirements for this one or is it OK to use the rook workflow as a start?
> 
> https://github.com/rook/rook/blob/master/.github/workflows/stale.yaml

I used that file as a basis after confirmation from @Madhu-1   on chat.

File: .github/workflows/stale.yaml
Comment (nb-ohad) [inline .github/workflows/stale.yaml:1]: Can we have a special label (or other mechanism) to enable us to keep alive a PR or an issue?

Comment (obnoxxx) [reply .github/workflows/stale.yaml:1]: @nb-ohad  wrote:
> Can we have a special label (or other mechanism) to enable us to keep alive a PR or an issue?

This is what the list of  exempt labels are for: `keepalive`etc.

We need to create the required labels in the project anyway. `stale`, `keepalive`etc.

Comment (nb-ohad) [reply .github/workflows/stale.yaml:1]: Thank you for clarifying, I missed that. I can create the labels after this PR is merged.

**Comment by obnoxxx** [comment]
marked ready for review since all items of #149 are addressed.

File: .github/workflows/mod-check.yaml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    branches:
+      - main
+      - release-*
+
+# cancel the in-progress workflow when PR is refreshed.
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event_name == 'pull_request' && github.head_ref || github.sha }}
+  cancel-in-progress: true
+
+permissions:
```
Comment (Madhu-1) [inline .github/workflows/mod-check.yaml:20]: as it's mostly around linter, adding it to https://github.com/ceph/ceph-csi-operator/blob/main/.github/workflows/lint.yaml makes more sense, WDYT?

File: Makefile [hunk 2]
```diff
@@ -109,6 +109,13 @@ fmt: ## Run go fmt against code.
 vet: ## Run go vet against code.
 	go vet ./...
 
+.PHONY: mod.check
+mod.check:#check go module dependencies
+	go mod verify
+	go mod tidy
```
Comment (Madhu-1) [inline Makefile:115]: This check should error out if there the check fails (i mean if there is a change in go.mod/go.sum/vendor) and can you please also check if we need to run it in  API package as we also have go.mod and go.sum over there as well.

**Review by Madhu-1 — CHANGES_REQUESTED** [review]

File: Makefile [hunk 3]
```diff
@@ -109,6 +109,13 @@ fmt: ## Run go fmt against code.
...
+
+
+
```
[the remark this answers is not in the document]
Comment (obnoxxx) [reply Makefile]: done

File: .github/workflows/mod-check.yaml (hunk 1 above)
Comment (obnoxxx) [reply .github/workflows/mod-check.yaml]: done. and removed the mod-check workflow

File: Makefile (hunk 2 above)
Comment (obnoxxx) [reply Makefile:115]: @Madhu-1  wrote:
> This check should error out if there the check fails (i mean if there is a change in go.mod/go.sum/vendor)

I don't quite get it yet: Why should there be a change there?
Do you mean that the operations like `go mod tidy` could lead to a change there and that should result in test failure?

>  and can you please also check if we need to run it in API package as we also have go.mod and go.sum over there as well.

sure, will do.

Comment (Madhu-1) [reply Makefile:115]: >I don't quite get it yet: Why should there be a change there?
Do you mean that the operations like go mod tidy could lead to a change there and that should result in test failure?

yes that's correct, that can result in a change, and file content will change, we have a similar check-in 

https://github.com/ceph/ceph-csi/blob/d1c28fa57a0319879dc516ade77c5ec0d90500e9/Makefile#L110-L114

File: .github/workflows/dependency-review.yaml [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+jobs:
+  dependency-review:
+    runs-on: ubuntu-latest
+    steps:
+      - name: 'Checkout Repository'
+        uses: actions/checkout@v4
+      - name: 'Dependency Review'
+        uses: actions/dependency-review-action@v4
+
```
[the remark this answers is not in the document]
Comment (obnoxxx) [reply .github/workflows/dependency-review.yaml]: done

File: .github/workflows/stale.yaml [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+permissions:
+  contents: read
+
+jobs:
+  stale:
+    permissions:
+      issues: write # for actions/stale to close stale issues
+      pull-requests: write # for actions/stale to close stale PRs
+    runs-on: ubuntu-22.04
+    if: github.repository == 'rook/rook'
```
[the remark this answers is not in the document]
Comment (obnoxxx) [reply .github/workflows/stale.yaml]: fixed, thanks!

File: .github/workflows/lint.yaml [hunk 6]
```diff
@@ -103,3 +103,18 @@ jobs:
         ignore_paths: 'vendor'
         disable_matcher: false
         format: gcc
+
```
Comment (Madhu-1) [inline .github/workflows/lint.yaml:106]: Can you please remove extra new line here

File: .github/workflows/lint.yaml [hunk 7]
```diff
@@ -103,3 +103,18 @@ jobs:
...
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 1
+      - uses: actions/setup-go@v5
+        with:
+          go-version-file: go.mod
+        name: run mod check
+         run: make mod.check
+
```
Comment (Madhu-1) [inline .github/workflows/lint.yaml:120]: Please remove new line here as well.

**Review by Madhu-1 — COMMENTED** [review]
small nits

File: .github/workflows/lint.yaml (hunk 6 above)
Comment (obnoxxx) [reply .github/workflows/lint.yaml:106]: done

File: .github/workflows/lint.yaml (hunk 7 above)
Comment (obnoxxx) [reply .github/workflows/lint.yaml]: done

**Review by Madhu-1 — APPROVED** [review]

**Pull request merged**